python/ray/serve/_private/autoscaling_state.py

```diff
@@ -753,14 +761,33 @@ def get_decision_num_replicas(
             # Using app-level policy
             autoscaling_contexts = {
                 deployment_id: state.get_autoscaling_context(
-                    deployment_to_target_num_replicas[deployment_id]
+                    deployment_to_target_num_replicas[deployment_id],
+                    self._policy_state.get(deployment_id, {})
+                    if self._policy_state
+                    else {},
                 )
                 for deployment_id, state in self._deployment_autoscaling_states.items()
             }
+            # Policy returns decisions: {deployment_id -> decision} and
+            # policy state: {deployment_id -> Dict}
+            decisions, returned_policy_state = self._policy(autoscaling_contexts)
 
-            # Policy returns {deployment_name -> decision}
-            decisions, self._policy_state = self._policy(autoscaling_contexts)
-
+            # Validate returned policy_state
+            if returned_policy_state is not None:
+                assert (
+                    type(returned_policy_state) is dict
+                ), "Application-level autoscaling policy must return policy_state as Dict[DeploymentID, Dict[str, Any]]"
+                # Check that all keys are valid deployment IDs
+                for deployment_id in returned_policy_state.keys():
+                    assert (
+                        deployment_id in self._deployment_autoscaling_states
+                    ), f"Policy state contains invalid deployment ID: {deployment_id}"
+                    assert (
+                        type(returned_policy_state[deployment_id]) is dict
+                    ), f"Policy state for deployment {deployment_id} must be a dictionary, got {type(returned_policy_state[deployment_id])}"
+            self._policy_state = returned_policy_state
```

Comment:
since this function is already to long, i suggest moving the validation into a helpful private function